R/bayes_factor.R

```diff
@@ -0,0 +1,96 @@
+#' Compute Bayes Factor (BF)
+#'
+#' Compute likelihood ratio between two competing hypotheses
+#'
+#' @param x Vector representing a posterior distribution
+#' @param method The method to be used. Currently only \code{"savage-dickey"} is supported
+#' @param ... Arguments passed to the specific methods
+#' @param prior Vector representing a prior distribution
+#' @param null Value to be tested
+#' @param direction Test type. One of \code{-1} (left tailed), \code{0} (defult; two tailed) or \code{1} (right tailed)
+#'
+#' @references
+#' Wagenmakers, E. J., Lodewyckx, T., Kuriyal, H., & Grasman, R. (2010). Bayesian
+#' hypothesis testing for psychologists: A tutorial on the Savage-Dickey method.
+#' Cognitive psychology, 60(3), 158-189.
+#'
+#' @author Mattan S. Ben-Shachar
+#'
+#' @export
+bayes_factor <- function(x,method = "savage-dickey",...){
+  UseMethod("bayes_factor")
+}
+
+#' @describeIn bayes_factor alias
+#' @export
+bf <- function(x,...) {
+  bayes_factor(x,...)
+}
+
+#' @rdname bayes_factor
+#' @export
+bayes_factor.numeric <- function(x,method = "savage-dickey",...){
+  method <- match.arg(method)
+  if (method=="savage-dickey") {
+    bf_val <- savage_dickey_bf(x,...)
+  } else {
+    stop("Other methods not supported yet.")
+  }
+
+  return(bf_val)
+}
+
+#' @rdname bayes_factor
+#' @export
+#' @import logspline
+savage_dickey_bf <- function(x,prior,null = 0,direction = 0){
+  if (require(logspline)) {
```

Comment:
I would change this into `requireNamespace()`, like

```r
if (!requireNamespace("logspline")) {
  stop("Package 'logspline' required.")
}
```

Reply:
Oops! That's what I meant to do!
Changed.

Reply:
Sorry, @strengejacke , you think this function should not run without `logspline`?
Note that the function can return a result using only `bayestestR` / `base` functions, but these tend to be less robust (though usually only slightly different) than those from `logspline`.

Reply:
We indeed try to keep mandatory dependencies to a minimum, which doesn't preclude having them in `suggests` as Daniel suggested. That means that people can install bayestestR without logspline, and use it, but if they want to use the function that uses it, they will be prompted with a message saying that they have to install it in order to use this function. If they already have it installed then no problem.

In general (we have this kind of stopping in several places throughout easystats), we should maybe reformulate the error message in a more informative and user-friendly manner, such as for example:

```r
stop("This function requires the package 'logspline' to be installed. Please install by running 'install.packages(logspline)' and run the function again :)") 
```

Reply:
I wasn't aware it also works w/o the logspline package. I'd suggest putting the code that does not rely on logspline into the if-block then, instead of the `stop()`. In any case, I think you should use `requireNamespace()` instead of `require()`, as this is what CRAN prefers.

Reply:
or simply replace `if (require(logspline)) {` by `if (requireNamespace("logspline")) {`

Reply:
Okay, so I added `logspline` in the suggest, and changed `if (require(logspline)) {` by `if (requireNamespace("logspline")) {`, and allowed the function to run without `logspline` with a warning.